### DISCUSSION TITLE
Error while onboarding External QR codes on Diksha

### DISCUSSION THREAD

**Discussion by samarthtrigyn** [Sunbird-Knowlg/Community#158] [General Discussions]
Hi @vinukumar-vs  @pramodkvarma @AmiableAnil As you might now on DIKSHA we are trying to onboard the QR codes generated on epaathshaala.
So in order to do that we have done the following steps:

1. Inserted the data in the Cassandra DB in the table dialcode_store> select * from dial_code where identifier='0125CH01';

identifier | batchcode                 | channel             | dialcode_index | generated_on            | metadata | published_on | publisher      | status
------------+---------------------------+---------------------+----------------+-------------------------+----------+--------------+----------------+--------
   0125CH01 | do_3138324509461872641811 | 0125196274181898243 |             -1 | 2023-10-15T10:00:00.000 |     null |         null | NCERTPUBLISHER |  Draft
*This was done mannually where the batchcode is the do_id of the content where we want to link the dialcode.

2. After the above step we tried to execute the api/dialcode/v1/content/link API to link the do Id with the request:: 
{
	"request": {
		"content": {
			"identifier": ["do_3138324509461872641811"],
			"dialcode": ["0125CH01"]
		}
	}
} 

But got the response:
{
    "id": "api.content.link",
    "ver": "1.0",
    "ts": "2023-10-18T06:34:53.531Z",
    "params": {
        "resmsgid": "72432ab0-6d80-11ee-ba46-cde7dc95c329",
        "msgid": null,
        "status": "failed",
        "err": "ERR_DIALCODE_LINK",
        "errmsg": "DIAL Code not found with id(s):[0125CH01]"
    },
    "responseCode": "RESOURCE_NOT_FOUND",
    "result": {}
}

We also get the error on the frontend while searching for the DIAL code as "Invalid code".

Can you please guide what are we missing, Im assuming that the syncing to the ES is not triggered as the dial reserve api is not called that's what's giving the error.
If that is the case how can we trigger the search indexer job mannually?

**Comment by pallakartheekreddy** [comment]
@samarthtrigyn You are correct. The data is not currently available in Elasticsearch for searching. To address this, please utilize the `Sync Tool` to synchronize the DIAL code metadata from Cassandra to Elasticsearch. Here are the specific details:

Please use `syncdialcodes` command by providing DIAL codes which we need to sync.

**Reply by samarthtrigyn** [reply]
Thanks for the quick response @pallakartheekreddy, can you guide me where do I need to run the syncdialcodes command? on which server? And where can I find the Sync Tool?

Thanks for all your help.

**Reply by pallakartheekreddy** [reply]
@samarthtrigyn `Sync Tool`is Jenkin's Job. Inside this job, there is an option labeled `Select a command.` Please choose the `syncdialcodes` option, as depicted in `Image 1`. After that, you can specify the dial codes that you wish to synchronize, as demonstrated in `Image 2`, and then proceed to execute the job.

**Image 1:**
<img width="257" alt="Screenshot 2023-10-19 at 11 15 11 AM" src="https://github.com/Sunbird-Knowlg/Community/assets/20140324/07540d94-e62e-47ca-9c23-999615ae0dfd">

**Image 2:**
<img width="1265" alt="Screenshot 2023-10-19 at 11 19 56 AM" src="https://github.com/Sunbird-Knowlg/Community/assets/20140324/35423ae3-f97b-4676-98e2-b987a0d16cba">

**Reply by Krishnaj20** [reply]
@samarthtrigyn - I hope your query is now resolved. As discussed, pls mark the thread as answered. Request you to share the resolution details so that wider community can benefit.

**Comment by samarthtrigyn** [comment]
Hi @Krishnaj20 and @pallakartheekreddy  Yes my query is closed thank you all for all the help.
On Diksha, the code for syncing the Cassandra DIALCODES is added in the jenkins job as shown below. 
<img width="244" alt="image" src="https://github.com/Sunbird-Knowlg/Community/assets/137397906/70f3968c-2f93-4847-8107-40a2a9b81442">


Also just for information to whosoever this might help, synctool is deployed on the lerning service VM and can be run mannually by going into the path "/home/learning/sync_tool" on the lerning service VM.

**Comment by samarthtrigyn** [comment]
Hi @pallakartheekreddy @Krishnaj20 @AmiableAnil,
I am facing a few other errors while onboarding the QR's.
So I have run the synctool and the QRs are synced, and as per me the next steps will be to execute the Content Link api for all the content DO_ids as mentioned below:

curl --location 'https://dev.oci.diksha.gov.in//api/dialcode/v1/content/link' \
--header 'X-Channel-ID: 0125196274181898243' \
--header 'Content-Type: application/json' \
--header 'Authorization:  \
--header 'x-authenticated-user-token:  \
--data '{
	"request": {
		"content": {
			"identifier": "do_3138714452083425281577",
			"dialcode": ["0122"]
		}
	}
}'

which gives me a success response 
{
    "id": "api.content.link",
    "ver": "1.0",
    "ts": "2023-10-23T11:53:11.249Z",
    "params": {
        "resmsgid": "bd746c10-719a-11ee-ba46-cde7dc95c329",
        "msgid": "bd65ed20-719a-11ee-ae50-e3ee8f092b94",
        "status": "successful",
        "err": null,
        "errmsg": null
    },
    "responseCode": "OK",
    "result": {}
}

But after running the above API suyccessfully I am still not ablew to search the content with the QR code.
Also do we need to link the parent textbook  do_Id with the same content api or we need to some other collection API?

Thanks 
Samarth Nigam

**Reply by pallakartheekreddy** [reply]
@samarthtrigyn Have you published the content after linking it to the dial code? Please share the content read API for reference.

Yes, for both collection and content, the same API can be used for linking to DIAL codes.
@Krishnaj20

**Reply by samarthtrigyn** [reply]
> Have you published the content after linking it to the dial code

Hi @pallakartheekreddy @Krishnaj20 I am not sure if it is linked or not because although the link api gives me a success response but no  node id or msg id is returned in the call as shown above. I have not published the content, I am trying to attach an external Dial code to already published and live content.

Please find the response of the content read call below:
curl --location 'https://dev.oci.diksha.gov.in//api/content/v1/read/do_3138714452083425281577' \
--header 'Content-Type: application/json' \
--header 'Authorization: '


{
    "id": "api.content.read",
    "ver": "1.0",
    "ts": "2023-10-25T09:33:45.854Z",
    "params": {
        "resmsgid": "981df1e0-7319-11ee-ba46-cde7dc95c329",
        "msgid": "981d2e90-7319-11ee-ae50-e3ee8f092b94",
        "status": "successful",
        "err": null,
        "errmsg": null
    },
    "responseCode": "OK",
    "result": {
        "content": {
            "ownershipType": [
                "createdBy"
            ],
            "copyright": "ncert",
            "se_gradeLevelIds": [
                "ncert_k-12_gradelevel_class1"
            ],
            "keywords": [
                "keu",
                "dxx",
                "testing",
                "kss",
                "key",
                "aug",
                "Tets",
                "Testibg"
            ],
            "year": "2008",
            "subject": [
                "Education"
            ],
            "channel": "0125196274181898243",
            "downloadUrl": "https://bmzbbujw9kal.compat.objectstorage.ap-mumbai-1.oraclecloud.com/dev-contents-storage/content/do_3138714452083425281577/aug29-qr-code_1693292134676_do_3138714452083425281577_1_SPINE.ecar",
            "organisation": [
                "NCERT"
            ],
            "language": [
                "English"
            ],
            "mimeType": "application/vnd.ekstep.content-collection",
            "variants": {
                "spine": {
                    "ecarUrl": "https://bmzbbujw9kal.compat.objectstorage.ap-mumbai-1.oraclecloud.com/dev-contents-storage/content/do_3138714452083425281577/aug29-qr-code_1693292134676_do_3138714452083425281577_1_SPINE.ecar",
                    "size": "92044"
                },
                "online": {
                    "ecarUrl": "https://bmzbbujw9kal.compat.objectstorage.ap-mumbai-1.oraclecloud.com/dev-contents-storage/content/do_3138714452083425281577/aug29-qr-code_1693292136481_do_3138714452083425281577_1_ONLINE.ecar",
                    "size": "17964"
                }
            },
            "leafNodes": [
                "do_313867301754281984110",
                "do_313867308528377856116",
                "do_31386730100358348819",
                "do_313867131273723904112",
                "do_313867162684489728115",
                "do_313867303391698944114",
                "do_3138686980589404161224",
                "do_313867156661092352114",
                "do_313867154417410048113",
                "do_31386730041211289618",
                "do_31386670003810304016",
                "do_31386669656661196815",
                "do_3138699994237583361443"
            ],
            "objectType": "Content",
            "se_mediums": [
                "English"
            ],
            "gradeLevel": [
                "Class 1"
            ],
            "appIcon": "https://bmzbbujw9kal.compat.objectstorage.ap-mumbai-1.oraclecloud.com/dev-contents-storage/collection/do_3138714452083425281577/artifact/_-__-b547u5p95ch-b2t4l5-5.thumb.-araa-ar3a-ar3arar-arar-ar2a-_1582352811346.thumb.jpg",
            "primaryCategory": "Digital Textbook",
            "appId": "dev.diksha.portal",
            "contentEncoding": "gzip",
            "lockKey": "1f2b6202-5426-4a44-8f5a-2468215b63fa",
            "generateDIALCodes": "Yes",
            "totalCompressedSize": 296471769,
            "mimeTypesCount": "{\"application/pdf\":4,\"video/mp4\":6,\"video/webm\":3,\"application/epub\":1,\"application/vnd.ekstep.content-collection\":16}",
            "sYS_INTERNAL_LAST_UPDATED_ON": "2023-08-29T06:54:15.355+0000",
            "contentType": "TextBook",
            "se_gradeLevels": [
                "Class 1"
            ],
            "trackable": {
                "enabled": "No",
                "autoBatch": "No"
            },
            "identifier": "do_3138714452083425281577",
            "audience": [
                "Student"
            ],
            "se_boardIds": [
                "ncert_k-12_board_cbse"
            ],
            "subjectIds": [
                "ncert_k-12_subject_education"
            ],
            "toc_url": "https://bmzbbujw9kal.compat.objectstorage.ap-mumbai-1.oraclecloud.com/dev-contents-storage/content/do_3138714452083425281577/artifact/do_3138714452083425281577_toc.json",
            "visibility": "Default",
            "contentTypesCount": "{\"LessonPlan\":1,\"TextBookUnit\":3,\"Collection\":11,\"LessonPlanUnit\":1,\"Resource\":14}",
            "author": "creator_ncert",
            "consumerId": "3482e7a5-5057-4da1-85f1-7150b2554582",
            "childNodes": [
                "do_3138707012190453761537",
                "do_3138714457025413121578",
                "do_313867162684489728115",
                "do_3138714459085783041580",
                "do_313856655379709952115",
                "do_3138714464695828481582",
                "do_31386730100358348819",
                "do_3138707018172006401550",
                "do_313867303391698944114",
                "do_3138707018172088321554",
                "do_3138686980589404161224",
                "do_313867156661092352114",
                "do_313867154417410048113",
                "do_3138707018171432961538",
                "do_3138707018172006401552",
                "do_313867301754281984110",
                "do_313867308528377856116",
                "do_313867131273723904112",
                "do_3138707018172088321556",
                "do_3138707018171842561544",
                "do_3138707018171760641542",
                "do_3138707018171678721540",
                "do_31386730041211289618",
                "do_3138707018171924481548",
                "do_31386670003810304016",
                "do_3138707018171842561546",
                "do_31386669656661196815",
                "do_3138699994237583361443",
                "do_313856655483985920116"
            ],
            "discussionForum": {
                "enabled": "No"
            },
            "mediaType": "content",
            "osId": "org.ekstep.quiz.app",
            "languageCode": [
                "en"
            ],
            "lastPublishedBy": "cf25f36e-f790-4c3c-a767-70a851e0dead",
            "version": 2,
            "se_subjects": [
                "Education"
            ],
            "license": "CC BY 4.0",
            "prevState": "Review",
            "qrCodeProcessId": "6aa89850-24b5-443c-9036-9b1a20f4af65",
            "size": 92044,
            "lastPublishedOn": "2023-08-29T06:55:34.302+0000",
            "name": "Aug29 qr code",
            "publisher": "jack",
            "mediumIds": [
                "ncert_k-12_medium_english"
            ],
            "status": "Live",
            "code": "org.sunbird.LQpijf",
            "credentials": {
                "enabled": "No"
            },
            "prevStatus": "Processing",
            "description": "Testing Purpsoe ",
            "medium": [
                "English"
            ],
            "posterImage": "https://obj.diksha.gov.in/ntp-content-production/content/do_3129626342270648321290/artifact/_-__-b547u5p95ch-b2t4l5-5.-araa-ar3a-ar3arar-arar-ar2a-_1582352811346.jpg",
            "idealScreenSize": "normal",
            "createdOn": "2023-08-29T06:47:31.414+0000",
            "reservedDialcodes": {
                "G1S1F6": 0,
                "U6E4Y5": 3,
                "K9Q5K6": 4,
                "P3N6S5": 2,
                "I4K4A9": 1
            },
            "se_boards": [
                "CBSE"
            ],
            "se_mediumIds": [
                "ncert_k-12_medium_english"
            ],
            "copyrightYear": 2023,
            "contentDisposition": "inline",
            "additionalCategories": [
                "Textbook"
            ],
            "lastUpdatedOn": "2023-08-29T06:55:36.601+0000",
            "dialcodeRequired": "Yes",
            "lastStatusChangedOn": "2023-08-29T06:55:36.601+0000",
            "createdFor": [
                "0125196274181898243"
            ],
            "creator": "creator_ncert",
            "os": [
                "All"
            ],
            "se_subjectIds": [
                "ncert_k-12_subject_education"
            ],
            "se_FWIds": [
                "ncert_k-12"
            ],
            "pkgVersion": 1,
            "versionKey": "1693292060874",
            "idealScreenDensity": "hdpi",
            "framework": "ncert_k-12",
            "dialcodes": [
                "0122"
            ],
            "depth": 0,
            "s3Key": "content/do_3138714452083425281577/artifact/do_3138714452083425281577_toc.json",
            "boardIds": [
                "ncert_k-12_board_cbse"
            ],
            "lastSubmittedOn": "2023-08-29T06:54:20.839+0000",
            "createdBy": "e2419c04-2862-44da-a1c2-28d9ab5d641f",
            "compatibilityLevel": 1,
            "leafNodesCount": 13,
            "userConsent": "Yes",
            "gradeLevelIds": [
                "ncert_k-12_gradelevel_class1"
            ],
            "board": "CBSE",
            "resourceType": "Book"
        }
    }
}

**Reply by pallakartheekreddy** [reply]
@samarthtrigyn Please publish the content again. Without publish, QR code search will not work.

**Reply by samarthtrigyn** [reply]
Hi @pallakartheekreddy if I add and publish the content again, then what is the the usecase of the Content Link API then. I assumed it is also used to change the DIAL codes in the content even when the content is already published and live. 
Also for publishing the content and the books again we will require the credentials of the same user that published the book in the first place right?

**Reply by pallakartheekreddy** [reply]
@samarthtrigyn The Content Link API is specifically used to link a DIAL code to a particular content. However, if you intend to perform a QR code search, it's necessary to publish the content. Even if the content is already published and live, you still need to follow the same process. Otherwise, the linking will apply only to the draft version of the content.

In that organization, any reviewer has the capability to publish content and books once more. There is no requirement to use the credentials of the same user who originally published the book.
@Krishnaj20

**Reply by samarthtrigyn** [reply]
Thanks @pallakartheekreddy this helps.

**Comment by samarthtrigyn** [comment]
HI @pallakartheekreddy, we are getting a strange issue. I ran the sync tool from the pipeline and are getting the error as shown below. We also tried to run the job manually on the server as well and get the same error. 
The same jar works on the dev environment. 

We also tried to run the job manually after removing the  Log4j2Plugins.dat from the jar that was built. That time there was no error on the console however the syncing did not happen and we were not able to search the synced dial code to attach in any book.

Can you point me in the right direction? What can we be missing?

ok: [10.90.14.21] => {
    "job_status": {
        "ansible_job_id": "470897293704.702683",
        "changed": true,
        "cmd": "java -Dconfig.file=/home/learning/sync_tool/application.conf -jar sync-tool-0.0.1-SNAPSHOT.jar syncbyids --ids 0122CH01",
        "delta": "0:00:02.050281",
        "end": "2023-10-26 12:32:33.858527",
        "failed": false,
        "finished": 1,
        "rc": 0,
        "start": "2023-10-26 12:32:31.808246",
        "stderr": "ERROR StatusLogger Unrecognized format specifier [d]\nERROR StatusLogger Unrecognized conversion specifier [d] starting at position 16 in conversion pattern.\nERROR StatusLogger Unrecognized format specifier [thread]\nERROR StatusLogger Unrecognized conversion specifier [thread] starting at position 25 in conversion pattern.\nERROR StatusLogger Unrecognized format specifier [level]\nERROR StatusLogger Unrecognized conversion specifier [level] starting at position 35 in conversion pattern.\nERROR StatusLogger Unrecognized format specifier [logger]\nERROR StatusLogger Unrecognized conversion specifier [logger] starting at position 47 in conversion pattern.\nERROR StatusLogger Unrecognized format specifier [msg]\nERROR StatusLogger Unrecognized conversion specifier [msg] starting at position 54 in conversion pattern.\nERROR StatusLogger Unrecognized format specifier [n]\nERROR StatusLogger Unrecognized conversion specifier [n] starting at position 56 in conversion pattern.\nERROR StatusLogger Unrecognized format specifier [d]\nERROR StatusLogger Unrecognized conversion specifier [d] starting at position 16 in conversion pattern.\nERROR StatusLogger Unrecognized format specifier [thread]\nERROR StatusLogger Unrecognized conversion specifier [thread] starting at position 25 in conversion pattern.\nERROR StatusLogger Unrecognized format specifier [level]\nERROR StatusLogger Unrecognized conversion specifier [level] starting at position 35 in conversion pattern.\nERROR StatusLogger Unrecognized format specifier [logger]\nERROR StatusLogger Unrecognized conversion specifier [logger] starting at position 47 in conversion pattern.\nERROR StatusLogger Unrecognized format specifier [msg]\nERROR StatusLogger Unrecognized conversion specifier [msg] starting at position 54 in conversion pattern.\nERROR StatusLogger Unrecognized format specifier [n]\nERROR StatusLogger Unrecognized conversion specifier [n] starting at position 56 in conversion pattern.",
        "stderr_lines": [
            "ERROR StatusLogger Unrecognized format specifier [d]",
            "ERROR StatusLogger Unrecognized conversion specifier [d] starting at position 16 in conversion pattern.",
            "ERROR StatusLogger Unrecognized format specifier [thread]",
            "ERROR StatusLogger Unrecognized conversion specifier [thread] starting at position 25 in conversion pattern.",
            "ERROR StatusLogger Unrecognized format specifier [level]",
            "ERROR StatusLogger Unrecognized conversion specifier [level] starting at position 35 in conversion pattern.",
            "ERROR StatusLogger Unrecognized format specifier [logger]",
            "ERROR StatusLogger Unrecognized conversion specifier [logger] starting at position 47 in conversion pattern.",
            "ERROR StatusLogger Unrecognized format specifier [msg]",
            "ERROR StatusLogger Unrecognized conversion specifier [msg] starting at position 54 in conversion pattern.",
            "ERROR StatusLogger Unrecognized format specifier [n]",
            "ERROR StatusLogger Unrecognized conversion specifier [n] starting at position 56 in conversion pattern.",
            "ERROR StatusLogger Unrecognized format specifier [d]",
            "ERROR StatusLogger Unrecognized conversion specifier [d] starting at position 16 in conversion pattern.",
            "ERROR StatusLogger Unrecognized format specifier [thread]",
            "ERROR StatusLogger Unrecognized conversion specifier [thread] starting at position 25 in conversion pattern.",
            "ERROR StatusLogger Unrecognized format specifier [level]",
            "ERROR StatusLogger Unrecognized conversion specifier [level] starting at position 35 in conversion pattern.",
            "ERROR StatusLogger Unrecognized format specifier [logger]",
            "ERROR StatusLogger Unrecognized conversion specifier [logger] starting at position 47 in conversion pattern.",
            "ERROR StatusLogger Unrecognized format specifier [msg]",
            "ERROR StatusLogger Unrecognized conversion specifier [msg] starting at position 54 in conversion pattern.",
            "ERROR StatusLogger Unrecognized format specifier [n]",
            "ERROR StatusLogger Unrecognized conversion specifier [n] starting at position 56 in conversion pattern."
        ],
        "stdout": "Total Number of Objects to be Synced : 1\nIdentifiers : [0122CH01]\n(1) Nodes not found: [0122CH01], remaining nodes got synced successfully\nTotal time of execution: 13ms\nSTART_TIME: 2023/10/26 12:32:33, END_TIME: 2023/10/26 12:32:33",
        "stdout_lines": [
            "Total Number of Objects to be Synced : 1",
            "Identifiers : [0122CH01]",
            "(1) Nodes not found: [0122CH01], remaining nodes got synced successfully",
            "Total time of execution: 13ms",
            "START_TIME: 2023/10/26 12:32:33, END_TIME: 2023/10/26 12:32:33"
        ]
    }
}
Thanks

**Reply by pallakartheekreddy** [reply]
@samarthtrigyn I'm pleased to hear that the issue has been resolved and is functioning correctly in the development environment. When reviewing above error, there are two aspects to consider:

1. Please ensure that the "0122CH01" dial code is available in the Cassandra DB. Follow the same process we executed in the development environment to verify its presence. If it's not in the database, please create it first and then attempt to synchronize.

2. We should verify the `application.conf` of the sync tool to confirm that it is pointing to the same keyspace and table where the aforementioned dial code exists.
@Krishnaj20

**Reply by samarthnigam** [reply]
Hi @pallakartheekreddy,
Yes we have verified that the Dial code is in the database in the correct keyspace and table.
We have also verified that the IP's configured in the application.conf is the correct one. Thats what is very surprising that the syncing is not happening.

**Reply by samarthnigam** [reply]
Hi @pallakartheekreddy @Krishnaj20 we have been able to resolve the issue and the QR codes are synced successfully on the platform. The issue that we were facing was the date time format inserted in to the dialcodestore was not in the correct format, and were getting the ElasticsearchException[Elasticsearch exception [type=illegal_argument_exception, reason=Invalid format: "2023-10-20T23:04:06:708+000" is malformed at ":708+000"]].
We inserted the date in the correct format as "2023-10-15T10:00:00.000" and the dialcodes were synced successfully.
One of the reasons it took us so much time to find the root cause was because in the Jenkins job the exception was not being logged on the console, rather it was being logged in the Telemetry API.

I would also Like to contribute to this finding in the code by adding some sysouts which can help someone facing the same issue.
So can you please guide me on how to go about it? can I just fork the repo, make the change that I did for this and request a PR? or I need to do something else for this as well.

**Reply by pallakartheekreddy** [reply]
@samarthnigam I am glad to hear that the issue has been resolved. Yes, Please go ahead and fork the repository, and then proceed to create a pull request(PR).

**Reply by Krishnaj20** [reply]
@samarthnigam - Super! We will be glad to have this enhancement as a contribution. Thank you!

@madhucr @vrayulu

**Reply by samarthnigam** [reply]
@pallakartheekreddy  @Krishnaj20 So for the PR Should I merge my changes into the master branch or any specific branch that I need to use? 
I created my branch from an older tag that we were using "release-5.2.0_RC3".

**Reply by pallakartheekreddy** [reply]
@samarthnigam Please send to `release-5.2.0` branch.

**Reply by samarthnigam** [reply]
@Krishnaj20  @pallakartheekreddy  PR added:: https://github.com/Sunbird-Knowlg/sunbird-learning-platform/pull/1977
